src/ray/raylet/scheduling_resources.cc

```diff
@@ -9,39 +9,106 @@ namespace ray {
-bool EqualsOneEpsilon(double quantity) {
-  if ((quantity <= 1 + EPSILON) && (quantity >= 1 - EPSILON)) {
-    return true;
-  }
-  return false;
+FractionalResourceQuantity::FractionalResourceQuantity(int resource_quantity) {
+  resource_name_ = "";
+  resource_quantity_ =
+      static_cast<int>(std::round(resource_quantity * conversion_factor_));
+}
+
+FractionalResourceQuantity::FractionalResourceQuantity(std::string resource_name,
+                                                       double resource_quantity) {
+  RAY_CHECK(resource_quantity > 0) << "Resource " << resource_name << " capacity is "
+                                   << resource_quantity
+                                   << ". Should have been greater than zero.";
+
+  resource_name_ = "";
+  resource_quantity_ =
+      static_cast<int>(std::round(resource_quantity * conversion_factor_));
+}
+
+const FractionalResourceQuantity FractionalResourceQuantity::operator+(
+    const FractionalResourceQuantity &rhs) const {
+  FractionalResourceQuantity result = *this;
+  result += rhs;
+  return result;
+}
+
+const FractionalResourceQuantity FractionalResourceQuantity::operator-(
+    const FractionalResourceQuantity &rhs) const {
+  FractionalResourceQuantity result = *this;
+  result -= rhs;
+  return result;
+}
+
+FractionalResourceQuantity &FractionalResourceQuantity::operator+=(
+    const FractionalResourceQuantity &rhs) {
+  resource_quantity_ += rhs.resource_quantity_;
+
+  return *this;
+}
+
+FractionalResourceQuantity &FractionalResourceQuantity::operator-=(
+    const FractionalResourceQuantity &rhs) {
+  resource_quantity_ -= rhs.resource_quantity_;
+
+  // Ensure that quantity is nonnegative.
+  RAY_CHECK(resource_quantity_ >= 0)
+      << "Capacity of resource " << resource_name_ << " after subtraction is negative ("
+      << this->ToDouble() << ")."
+      << " Debug: resource_capacity_: (" << resource_name_ << ", " << this->ToDouble()
+      << "), other: (" << rhs.resource_name_ << ", " << rhs.ToDouble() << ").";
+  return *this;
+}
+
+bool FractionalResourceQuantity::operator==(const FractionalResourceQuantity &rhs) const {
+  return resource_quantity_ == rhs.resource_quantity_;
+}
+
+bool FractionalResourceQuantity::operator!=(const FractionalResourceQuantity &rhs) const {
+  return !(*this == rhs);
+}
+
+bool FractionalResourceQuantity::operator<(const FractionalResourceQuantity &rhs) const {
+  return resource_quantity_ < rhs.resource_quantity_;
+}
+
+bool FractionalResourceQuantity::operator>(const FractionalResourceQuantity &rhs) const {
+  return rhs < *this;
+}
+
+bool FractionalResourceQuantity::operator<=(const FractionalResourceQuantity &rhs) const {
+  return !(*this > rhs);
+}
+
+bool FractionalResourceQuantity::operator>=(const FractionalResourceQuantity &rhs) const {
+  bool result = !(*this < rhs);
+  return result;
+}
+
+double FractionalResourceQuantity::ToDouble() const {
+  return static_cast<double>(resource_quantity_ / conversion_factor_);
```

Comment:
Won't this always return 0? This looks like a bug.

Reply:
This suggests that we need to augment the test to catch this bug.

Reply:
Would a bug like that be caught in `test_basic.py::test_resource_assignment`?
https://github.com/ray-project/ray/blob/1592ead002086768b8ec6ea609bd9a625a1dc08d/python/ray/tests/test_actor.py#L560-L561